File: api/lib/routes/routeHelper.js
'use strict';

var DESC = exports.DESC = "DESC";
var ASC = exports.ASC = "ASC";

var isMissing = exports.isMissing = function(value) {
  return typeof(value) === "undefined" || value === null;
}
var isInteger = exports.isInteger = function(value) {
  return Number.isInteger(value);
}
var isString = exports.isString = function(value) {
  return typeof(value) === "string";
}

exports.parseParameter = function(req) {
  var parameters = {};
  var appId = req.params.app_id;
  var startTime = req.query["start-time"];
  var endTime = req.query["end-time"];
  var orderDirection = req.query["order-direction"] ? req.query["order-direction"] : req.query["order"];
  var page = req.query["page"];
  var resultsPerPage = req.query["results-per-page"];

  if (!appId) {
    return { valid: false, message: "app_id is required" };
  } else {
    parameters.appId = appId;
  }
  if (isMissing(startTime)) {
    parameters.startTime = 0;
  } else {
    startTime = parseInt(startTime);
    if (!isInteger(startTime)) {
      return { valid: false, message: "start-time must be an integer" };
    } else {
      parameters.startTime = startTime;
    }
  }

  if (isMissing(endTime)) {
    parameters.endTime = -1;
  } else {
    endTime = parseInt(endTime);
    if (!isInteger(endTime)) {
      return { valid: false, message: "end-time must be an integer" };
    } else {
      parameters.endTime = endTime;
    }
  }
  if (isMissing(orderDirection)) {
    parameters["order"] = DESC;
  } else {
    if (!isString(orderDirection)) {
      return { valid: false, message: "order-direction must be a string" };
    } else {
      orderDirection = orderDirection.toUpperCase();
      if (orderDirection !== DESC && orderDirection !== ASC) {
        return { valid: false, message: "order-direction must be DESC or ASC" };
      } else {
        parameters["order"] = orderDirection;
      }
    }
  }
  if (isMissing(page)) {
    parameters.page = 1;
  } else {
    page = parseInt(page);
    if (!isInteger(page)) {
      return { valid: false, message: "page must be an integer" };
    }
    if (page <= 0) {
      return { valid: false, message: "page must be greater than 0" };
    }
    parameters.page = page;
  }

  if (isMissing(resultsPerPage)) {
    parameters.resultsPerPage = 50;
  } else {
    resultsPerPage = parseInt(resultsPerPage)
    if (!isInteger(resultsPerPage)) {
      return { valid: false, message: "results-per-page must be an integer" };
    }
    if (resultsPerPage <= 0) {
      return { valid: false, message: "results-per-page must be greater than 0" };
    }
    parameters.resultsPerPage = resultsPerPage;
  }


  return { valid: true, parameters: parameters }
}

exports.pagination = function(resultList, page, resultsPerPage, req) {
  var result = {};
  var totalResults = resultList.length;
  var totalPages = 0;
  var perPage = Number.parseInt(resultsPerPage);
  var pageNo = Number.parseInt(page);
  var prevUrl = null;
  var nextUrl = null;
  totalPages = (totalResults % perPage) === 0 ? (totalResults / perPage) : (Number.parseInt(totalResults / perPage) + 1);
  var startIndex = (pageNo - 1) * perPage;
  var endIndex = startIndex + perPage;
  var resources = resultList.slice(startIndex, endIndex);
  var originalUrl = req.originalUrl
  if (page > 1 && page <= totalPages + 1) {
    var prevPage = page - 1;
    prevUrl = getPageUrl(req, prevPage);
  }
  if (page < totalPages) {
    var nextPage = page + 1;
    nextUrl = getPageUrl(req, nextPage);
  }
  result.total_results = totalResults;
  result.total_pages = Number.parseInt(totalPages);
  result.page = Number.parseInt(pageNo);
  result.prev_url = prevUrl;
  result.next_url = nextUrl;
  result.resources = resources;
  return result;
}

var getPageUrl = exports.getPageUrl = function(req, targetPageNo) {
  var queries = req.query;
  if (!queries || !queries["page"]) {
    return null;
  }
  queries["page"] = targetPageNo;
  var pageStr = req.baseUrl + req.path + "?";
  for (var key in queries) {
    pageStr += "&" + key + "=" + queries[key];
  }
  pageStr = pageStr.replace("&", "");
  return pageStr;
}

File: api/publicapiserver/route_helper.go
package publicapiserver

import (
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"strconv"
	"strings"

	"code.cloudfoundry.org/app-autoscaler/src/autoscaler/models"
)

const (
	ASC  = "ASC"
	DESC = "DESC"
)

func parseParameter(r *http.Request, vars map[string]string) (*url.Values, error) {
	appId := vars["appId"]
	startTime := r.URL.Query().Get("start-time")
	endTime := r.URL.Query().Get("end-time")
	orderDirection := r.URL.Query().Get("order-direction")
	order := r.URL.Query().Get("order")
	page := r.URL.Query().Get("page")
	resultsPerPage := r.URL.Query().Get("results-per-page")

	if appId == "" {
		return nil, fmt.Errorf("appId is required")
	}

	if startTime == "" {
		startTime = "0"
	}
	_, err := strconv.Atoi(startTime)
	if err != nil {
		return nil, fmt.Errorf("start-time must be an integer")
	}

	if endTime == "" {
		endTime = "-1"
	}
	_, err = strconv.Atoi(endTime)
	if err != nil {
		return nil, fmt.Errorf("end-time must be an integer")
	}

	if orderDirection == "" && order == "" {
		orderDirection = DESC
	} else if orderDirection == "" && order != "" {
		orderDirection = order
	}
	orderDirection = strings.ToUpper(orderDirection)
	if orderDirection != DESC && orderDirection != ASC {
		return nil, fmt.Errorf("order-direction must be DESC or ASC")
	}
	if page == "" {
		page = "1"
	}
	pageNo, err := strconv.Atoi(page)
	if err != nil {
		return nil, fmt.Errorf("page must be an integer")
	}
	if pageNo <= 0 {
		return nil, fmt.Errorf("page must be greater than 0")
	}

	if resultsPerPage == "" {
		resultsPerPage = "50"
	}
	resultsPerPageCount, err := strconv.Atoi(resultsPerPage)
	if err != nil {
		return nil, fmt.Errorf("results-per-page must be an integer")
	}
	if resultsPerPageCount <= 0 {
		return nil, fmt.Errorf("results-per-page must be greater than 0")
	}
	parameters := &url.Values{}
	parameters.Add("start", startTime)
	parameters.Add("end", endTime)
	parameters.Add("order", orderDirection)
	parameters.Add("page", page)
	parameters.Add("results-per-page", resultsPerPage)

	return parameters, nil
}

func paginateResource(resourceList []byte, parameters *url.Values, r *url.URL) (interface{}, error) {
	var resourceListItems []interface{}

	err := json.Unmarshal(resourceList, &resourceListItems)
	if err != nil {
		return nil, fmt.Errorf("failed to unmarshal resources")
	}
	totalResults := len(resourceListItems)
	perPage, _ := strconv.Atoi(parameters.Get("results-per-page"))
	pageNo, _ := strconv.Atoi(parameters.Get("page"))

	totalPages := 0
	if (totalResults % perPage) == 0 {
		totalPages = totalResults / perPage
	} else {
		totalPages = totalResults/perPage + 1
	}

	startIndex := (pageNo - 1) * perPage
	if startIndex > totalResults {
		startIndex = totalResults
	}
	endIndex := startIndex + perPage
	if endIndex > totalResults {
		endIndex = totalResults
	}

	resources := resourceListItems[startIndex:endIndex]
	prevUrl := ""
	if (pageNo > 1) && (pageNo <= totalPages+1) {
		prevUrl = getPageUrl(r, pageNo-1)
	}

	nextUrl := ""
	if pageNo < totalPages {
		nextUrl = getPageUrl(r, pageNo+1)
	}

	result := models.PublicApiResponseBase{}

	result.TotalResults = totalResults
	result.TotalPages = totalPages
	result.Page = pageNo
	result.PrevUrl = prevUrl
	result.NextUrl = nextUrl
	result.Resources = resources

	return result, nil
}

func getPageUrl(r *url.URL, targetPageNo int) string {
	pageUrl := *r
	queries := pageUrl.Query()
	pageParams := url.Values{}
	for key, value := range queries {
		if key == "page" {
			pageParams.Add(key, strconv.Itoa(targetPageNo))
		} else {
			pageParams.Add(key, value[0])
		}
	}

	pageUrl.RawQuery = pageParams.Encode()
	return pageUrl.String()
}
